src/BinSkim.Rules/PERules/BA6001.EnableSourceLink.cs

```diff
@@ -0,0 +1,88 @@
+using Microsoft.CodeAnalysis.IL.Sdk;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+
+namespace Microsoft.CodeAnalysis.IL.Rules
+{
+    [Export(typeof(Skimmer<BinaryAnalyzerContext>)), Export(typeof(ReportingDescriptor)), Export(typeof(IOptionsProvider))]
+    public class EnableSourceLink : WindowsBinaryAndPdbSkimmerBase, IOptionsProvider
+    {
+        /// <summary>
+        /// BA6001
+        /// </summary>
+        public override string Id => RuleIds.EnableSourceLink;
+
+        /// <summary>
+        /// Enable SourceLink.
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.BA6001_EnableSourceLink_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.BA6001_Pass),
+            nameof(RuleResources.BA6001_Error)
+        };
+
+        private const string AnalyzerName = RuleIds.EnableSourceLink + "." + nameof(EnableSourceLink);
+
+        public override void Initialize(BinaryAnalyzerContext context)
+        {
+            base.Initialize(context);
+        }
+
+        public override AnalysisApplicability CanAnalyzePE(PEBinary target, PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void AnalyzePortableExecutableAndPdb(BinaryAnalyzerContext context)
+        {
+            if (HasSourceLink(context))
+            {
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                    nameof(RuleResources.BA6001_Pass),
+                    context.TargetUri.GetFileName()));
+            }
+            else
+            {
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.Fail, context, null,
+                    nameof(RuleResources.BA6001_Error),
+                    context.TargetUri.GetFileName()));
+            }
+        }
+
+        private static bool HasSourceLink(BinaryAnalyzerContext context)
+        {
+            PEBinary target = context.PEBinary();
+            Pdb pdb = target.Pdb;
+
+            // We're just checking for the presence of SourceLink document(s),
+            // not whether they can be read.
+            if (pdb.FileType == PdbFileType.Portable)
+            {
+                string sourceLinkDocument = target.PE.ManagedPdbGetSourceLinkDocument(pdb);
+                return !string.IsNullOrEmpty(sourceLinkDocument);
+            }
+            else
+            {
+                IEnumerable<string> sourceLinkDocuments = pdb.WindowsPdbGetSourceLinkDocuments();
+                return sourceLinkDocuments != null && sourceLinkDocuments.Any();
```

Comment:
>Any [](http://example.com/codeflow?start=74&length=3)

Looking at your yield iterator implementation, looks like this code path is as efficient as it could be, i.e., breaks on the first evidence we've got support.

Just asking: is there no magic value, other (reliable) heuristic we can use to make this determination? I ask merely because as a rule, we try to avoid PDB parsing (and COM interop) wherever possible. Sniffing a magic file in a binary is admirably low-cost. :)

My sense is there's no opportunity here and we've got a reasonably efficient implementation (save the unnecessary presence of ReadUTF8 in the Portable code path, if I'm correct there's a fix we could make here).

Reply:
save the unnecessary presence of ReadUTF8 in the Portable code path ------ read form the test file i can see the linker commandline have the:
`/sourcelink`

 /ERRORREPORT:PROMPT /OUT:V:\T\binskim\src\BuildSamples\SourceLink\x64\Release\CPlusPlus_SourceLink.exe /NOLOGO /MANIFEST "/MANIFESTUAC:level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:V:\T\binskim\src\BuildSamples\SourceLink\x64\Release\CPlusPlus_SourceLink.pdb /SUBSYSTEM:CONSOLE /OPT:REF /OPT:ICF /LTCG:incremental /LTCGOUT:x64\Release\CPlusPlus_SourceLink.iobj /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:V:\T\binskim\src\BuildSamples\SourceLink\x64\Release\CPlusPlus_SourceLink.lib /MACHINE:X64 **/sourcelink:x64\Release\CPlusPlus_SourceLink.sourcelink.json**

[https://docs.microsoft.com/en-us/cpp/build/reference/sourcelink?view=msvc-170](https://docs.microsoft.com/en-us/cpp/build/reference/sourcelink?view=msvc-170)